### DISCUSSION TITLE
[Python] Is it possible to dynamically change proxy while scraping?

### DISCUSSION THREAD

**Discussion by mcharles96** [ultrafunkamsterdam/undetected-chromedriver#1267] [Q&A]
Hello guys!

I have a few codes that scrapes some websites running and working, but neither of that codes have a proxy setup. I tried some approaches over the internet to use public proxy to access websites, but my IP still being detected.

Is that possible to use a free proxy list without username and password authentication to make the requests? I'm currently using the _undetected-chromedriver_ to do the tasks.

Thank you!

**Comment by ManiMozaffar** [comment]
You've to do that on your network, that's the best way of doing it.

Also you can do using extension.
By default, if you bound chrome to a proxy, then it keep using that.

**Reply by mcharles96** [reply]
But I have two problems/doubts:
1. And if the website blocks me or doesn't allow web scraping?
2. Assuming that 1 is true, how can I use proxy inside cloud? I'm using Databricks today to do the scrapes.

**Comment by Charlotte-br560** [comment]
Yes, you can switch proxies dynamically. Try using libraries like `Requests` or `ProxyBroker` for easy proxy management. Also, consider checking out [Crawlbase](https://crawlbase.com/) for reliable proxies.

**Comment by st681** [comment]
Free proxies will be banned everywhere. You can use static/rotating residential proxies from many different providers for about 1USD/GB.

**Reply by denixyd** [reply]
can you refer to a website that provides quality residential proxies

**Comment by peipeiwei303-netizen** [comment]
Yes, dynamic proxy switching is possible, but it depends on what you mean by “dynamic”.

There are usually two patterns:

1. Per-request rotation (stateless scraping)
   This is common with requests-style workflows. Each HTTP call can use a different proxy, but you lose session continuity (cookies/login state won’t persist naturally).

2. Sticky session proxying (stateful scraping)
   Here you keep the same proxy for a period of time or session ID, which is essential for:

   login flows
   Selenium / Playwright sessions
   any site tracking cookies or session tokens

The key difference is:

   rotation = anonymity / distribution
   sticky = behavioral consistency

In Selenium or undetected-chromedriver setups, proxy switching is trickier because the proxy is usually bound at browser startup. So “dynamic switching” often means restarting the browser context with a new proxy rather than swapping it mid-session.

In practice, tools like Novada are often used in these cases because they support both rotating and sticky residential sessions, which makes it easier to align proxy behavior with browser automation without breaking session state.
